pkg/apis/machine/v1beta1/machine_webhook.go

```diff
@@ -0,0 +1,330 @@
+			return true, nil
+		}
+	}
+	return h
+}
+
+// NewDefaulter returns a new defaulterHandler.
+func NewMachineDefaulter() (*defaulterHandler, error) {
+	infra, err := getInfra()
+	if err != nil {
+		return nil, err
+	}
+
+	return createMachineDefaulter(infra.Status.PlatformStatus.Type, infra.Status.InfrastructureName), nil
+}
+
+func createMachineDefaulter(platform osconfigv1.PlatformType, clusterID string) *defaulterHandler {
+	h := &defaulterHandler{
+		clusterID: clusterID,
+	}
+
+	switch platform {
+	case osconfigv1.AWSPlatformType:
+		h.webhookOperations = defaultAWS
+	default:
+		// just no-op
+		h.webhookOperations = func(h *defaulterHandler, m *Machine) (bool, utilerrors.Aggregate) {
+			return true, nil
+		}
+	}
+	return h
+}
+
+// InjectDecoder injects the decoder.
+func (v *validatorHandler) InjectDecoder(d *admission.Decoder) error {
+	v.decoder = d
+	return nil
+}
+
+// InjectDecoder injects the decoder.
+func (v *defaulterHandler) InjectDecoder(d *admission.Decoder) error {
+	v.decoder = d
+	return nil
+}
+
+// Handle handles HTTP requests for admission webhook servers.
+func (h *validatorHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	m := &Machine{}
+
+	if err := h.decoder.Decode(req, m); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	klog.V(3).Infof("Validate webhook called for Machine: %s", m.GetName())
+
+	if ok, err := h.webhookOperations(h, m); !ok {
+		return admission.Denied(err.Error())
+	}
+
+	return admission.Allowed("Machine valid")
+}
+
+// Handle handles HTTP requests for admission webhook servers.
+func (h *defaulterHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	m := &Machine{}
+
+	if err := h.decoder.Decode(req, m); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	klog.V(3).Infof("Mutate webhook called for Machine: %s", m.GetName())
+
+	if ok, err := h.webhookOperations(h, m); !ok {
+		return admission.Denied(err.Error())
+	}
+
+	marshaledMachine, err := json.Marshal(m)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledMachine)
+}
+
+func defaultAWS(h *defaulterHandler, m *Machine) (bool, utilerrors.Aggregate) {
+	klog.V(3).Infof("Defaulting AWS providerSpec")
+
+	var errs []error
+	providerSpec := new(aws.AWSMachineProviderConfig)
+	if err := yaml.Unmarshal(m.Spec.ProviderSpec.Value.Raw, &providerSpec); err != nil {
+		errs = append(
+			errs,
+			field.Invalid(
+				field.NewPath("providerSpec", "value"),
+				providerSpec,
+				err.Error(),
+			),
+		)
+		return false, utilerrors.NewAggregate(errs)
+	}
+
+	if m.GetLabels() == nil {
+		m.Labels = make(map[string]string)
+	}
+	m.Labels["machine.openshift.io/cluster-api-cluster"] = h.clusterID
+
+	if providerSpec.InstanceType == "" {
+		providerSpec.InstanceType = "m4.large"
```

Comment:
i wonder if we should put these defaults into a big block of constants?

and by "these" i mean this one and the following as well. just thinking out loud here.

Reply:
done, updated here and the tests.